action/protocol/poll/governance_protocol.go

```diff
@@ -268,23 +307,29 @@ func (p *governanceChainCommitteeProtocol) readActiveBlockProducersByEpoch(ctx c
 		blockProducerList = append(blockProducerList, bp.Address)
 		blockProducerMap[bp.Address] = bp
 	}
-	bcCtx := protocol.MustGetBlockchainCtx(ctx)
-	rp := rolldpos.MustGetProtocol(bcCtx.Registry)
-	epochHeight := rp.GetEpochHeight(epochNum)
-	crypto.SortCandidates(blockProducerList, epochHeight, crypto.CryptoSeed)
 
-	// TODO: kick-out unqualified delegates based on productivity
+	crypto.SortCandidates(blockProducerList, epochHeight, crypto.CryptoSeed)
 
 	length := int(p.numDelegates)
-	if len(blockProducerList) < int(p.numDelegates) {
+	if len(blockProducerList) < length {
 		// TODO: if the number of delegates is smaller than expected, should it return error or not?
 		length = len(blockProducerList)
+		log.L().Warn(
+			"the number of block producer is less than expected",
+			zap.Int("actual block producer", len(blockProducerList)),
+			zap.Uint64("expected", p.numDelegates),
+		)
 	}
-
 	var activeBlockProducers state.CandidateList
 	for i := 0; i < length; i++ {
 		activeBlockProducers = append(activeBlockProducers, blockProducerMap[blockProducerList[i]])
 	}
+	// write into DB
+	delegateKey := candidatesutil.ConstructDelegateKey(epochNum)
+	if err := p.sm.PutState(delegateKey, &activeBlockProducers); err != nil {
+		return nil, err
+	}
```

Comment:
this is a read function, so we shouldn't do sm.PutState here. Else, something weird will happen here: https://github.com/iotexproject/iotex-core/blob/d31309f628d34aea9d941026941f4a6a47ff4daa/action/protocol/poll/governance_protocol.go#L208

Back to the solution: do we need to store delegates into state db?

Reply:
after off-line discussion, decide to store blacklist delegate into state DB through createPreState()